### PR TITLE
Add kinetic integral for HF solver

### PR DESCRIPTION
**Pull request by soranjh**
**Context:**
This PR adds functions to compute the kinetic integral between primitive or contracted Gaussian functions.

**Description of the Change:**
Functions `generate_kinetic` and `gaussian_kinetic` compute the kinetic integral.
Function `_diff2` computes the second order differentiated integral needed for evaluating a kinetic integral.

**Benefits:**
These functions are necessary for computing the kinetic integral over basis functions which is a requirement to perform Hartree-Fock calculation.

**Possible Drawbacks:**
No specific drawbacks.

**Related GitHub Issues:**
No related issues.

### CODE REVIEW

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:313]:
```suggestion
    The second-order integral :math:`D_{ij}^2`, where :math:`i` and :math:`j` denote angular
```

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:326]:
```suggestion
        rj (float): position component of the the second Gaussian function
```

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:331]:
```suggestion
        array[float]: second-order differentiated integral between two Gaussian functions
```

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:343]:
```suggestion
    r"""Compute the kinetic integral for two primitive Gaussian functions.
```

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:380]: Amazing docstring!

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:427]: What happens if you just use
``` 
generate_kinetic(mol.basis_set[0], mol.basis_set[1])()
```
?
Do we really need to pass an empty list of args?

File: pennylane/hf/integrals.py
Comment (ixfoduap) [inline pennylane/hf/integrals.py:432]: Any reason not to include a mathematical formula and example code as in the previous cases?

I think it would be best to include them here too

File: tests/hf/test_integrals.py
Comment (ixfoduap) [inline tests/hf/test_integrals.py:316]: Same comment as above

File: tests/hf/test_integrals.py
Comment (ixfoduap) [inline tests/hf/test_integrals.py:323]: This is zero because the H atoms are very far apart? It may be helpful to specify why this is a trivial case

File: tests/hf/test_integrals.py
Comment (ixfoduap) [inline tests/hf/test_integrals.py:384]: I know it's annoying, but should we be testing derivatives with respect to geometry as well?

Does it help to use pennylane's `finite_diff` function?

File: tests/hf/test_integrals.py
Comment (ixfoduap) [inline tests/hf/test_integrals.py:297]: My understanding of this test is that it checks if diff_2 is zero for functions with the same center, right? Better to specify this in the description, which is currently too general

**Review by ixfoduap — CHANGES_REQUESTED** [review]
PR looks great!

Just a few minor requests:
- Should probably add math and example use to the `kinetic_integral` docstrings
- Should probably test derivatives with respect to geometry too

File: tests/hf/test_integrals.py
Comment (soranjh) [reply tests/hf/test_integrals.py:384]: Added gradient test for `r`. I preferred to generate the reference gradients without using another function, but I can use `finite_diff` as well. I do not see any specific advantage except simplifying the code a bit, in the expense of adding an external function to the test function.

File: tests/hf/test_integrals.py
Comment (soranjh) [reply tests/hf/test_integrals.py:323]: Yes, explained it.

File: tests/hf/test_integrals.py
Comment (soranjh) [reply tests/hf/test_integrals.py:316]: Added a description. The functions are far from each other and the integral is zero.

File: tests/hf/test_integrals.py
Comment (soranjh) [reply tests/hf/test_integrals.py:297]: No, actually it returns zero when the functions are far from each other. Added a description.

File: pennylane/hf/integrals.py
Comment (soranjh) [reply pennylane/hf/integrals.py:427]: No, we do not, `generate_kinetic(mol.basis_set[0], mol.basis_set[1])()` will work too. However, I changed the example to a case where we have a differentiable parameter.

File: pennylane/hf/integrals.py
Comment (soranjh) [reply pennylane/hf/integrals.py:434]: This function is not visible to the users because it has to be used with `generate_kinetic`, which passes the basis functions to `kinetic_integral`. Therefore, we do not really need an example for this function, but I can copy the equations if they are helpful.

**Review by ixfoduap — APPROVED** [review]
Let's wait for @rmoyard's review before merging, but looks good to me!

**Review by rmoyard — COMMENTED** [review]
Great PR and tests, I've left some docstrings comments 💯

File: tests/hf/test_integrals.py
Comment (agran2018) [inline tests/hf/test_integrals.py:285]: Do we have a number to compare with for the case of non-zero overlap? If yes, I would suggest to add it.

**Review by agran2018 — APPROVED** [review]
Thank you @soranjh for the nice PR!. Looks good on my side.
My main comment is that based on the docstring of the ``_diff2`` function, it may be good to add an internal comment that ``_diif2`` can currently handle s- and p-type  GTOs.  If the implemented function supports d-type GTOs, then probably is just a matter of updating the docstring.

**Pull request merged**
